lib/Dialect/RTG/Transforms/ElaborationPass.cpp

```diff
@@ -0,0 +1,549 @@
+// for each invokation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/RTG/IR/RTGOps.h"
+#include "circt/Dialect/RTG/IR/RTGVisitors.h"
+#include "circt/Dialect/RTG/Transforms/RTGPasses.h"
+#include "mlir/IR/PatternMatch.h"
+#include <deque>
+#include <random>
+
+namespace circt {
+namespace rtg {
+#define GEN_PASS_DEF_ELABORATIONPASS
+#include "circt/Dialect/RTG/Transforms/RTGPasses.h.inc"
+} // namespace rtg
+} // namespace circt
+
+using namespace mlir;
+using namespace circt;
+using namespace circt::rtg;
+
+#define DEBUG_TYPE "rtg-elaboration"
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Elaborator Values
+//===----------------------------------------------------------------------===//
+
+/// The base class for elaborated values. Using the base class directly
+/// represents an opaque value, i.e., an SSA value which we cannot further
+/// interpret.
+/// Derived classes will also hold the opaque value such that we always have an
+/// SSA value at hand that we can use as a replacement for the concrete
+/// interpreted value when needed (alternatively we could materialize new IR,
+/// but that's more expensive).
+/// The derived classes are supposed to
+/// * add fields necessary to hold the concrete value
+/// * override the virtual methods to compute equivalence based on the
+///   interpreted values
+/// * implement LLVM's RTTI mechanism
+class ElaboratorValue {
+public:
+  /// Creates an opaque value.
+  ElaboratorValue(Value value) : ElaboratorValue(value, true) {
+    assert(value && "null values not allowed");
+  }
+
+  virtual ~ElaboratorValue() {}
+
+  Type getType() const { return value.getType(); }
+  Value getOpaqueValue() const { return value; }
+  bool isOpaqueValue() const { return isOpaque; }
+  virtual bool containsOpaqueValue() const { return isOpaque; }
+
+  virtual llvm::hash_code getHashValue() const {
+    return llvm::hash_combine(value, isOpaque);
+  }
+
+  virtual bool isEqual(const ElaboratorValue &other) const {
+    return isOpaque == other.isOpaque && value == other.value;
+  }
+
+protected:
+  ElaboratorValue(Value value, bool isOpaque)
+      : isOpaque(isOpaque), value(value) {}
+
+private:
+  const bool isOpaque;
+  const Value value;
+};
+
+/// Holds an evaluated value of a `SetType`'d value.
+class SetValue : public ElaboratorValue {
+public:
+  SetValue(Value value, SmallVector<ElaboratorValue *> &&set,
+           bool debug = false)
+      : ElaboratorValue(value, false), originalSize(set.size()), debug(debug),
+        set(set) {
+    assert(isa<SetType>(value.getType()));
+
+    if (debug)
+      for (auto [i, val] : llvm::enumerate(set))
+        debugMap[i] = val;
+
+    // Make sure the vector is sorted and has no duplicates.
+    llvm::sort(this->set);
+    this->set.erase(std::unique(this->set.begin(), this->set.end()),
+                    this->set.end());
+  }
```

Comment:
SmallVector has a single-iterator version of erase: `this->set.erase(llvm::unique(this->set))`